syntax = "proto3";
package org.solana.sealevel.v1;

// A set of feature flags.
message FeatureSet {
  // Every item in this list marks an enabled feature.  The value of
  // each item is the first 8 bytes of the feature ID as a little-
  // endian integer.
  repeated fixed64 features = 1;
}

// The complete state of an account excluding its public key.
message AcctState {
  // The account address.  (32 bytes)
  bytes address = 1;

  uint64 lamports = 2;

  // Account data is limited to 10 MiB on Solana mainnet as of 2024-Feb.
  bytes data = 3;

  bool executable = 4;

  // The rent epoch is deprecated on Solana mainnet as of 2024-Feb.
  // If ommitted, implies a value of UINT64_MAX.
  uint64 rent_epoch = 5;

  // Address of the program that owns this account.  (32 bytes)
  bytes owner = 6;
}

// EpochContext includes context scoped to an epoch.
// On "real" ledgers, it is created during the epoch boundary.
message EpochContext {
  FeatureSet features = 1;
}

// SlotContext includes context scoped to a block.
// On "real" ledgers, it is created during the slot boundary.
message SlotContext {}

// TODO: Include return data within the transaction context.
message TxnContext {}

message InstrAcct {
  // Selects an account in an external list
  uint32 index = 1;
  bool is_writable = 2;
  bool is_signer = 3;
}

// The execution context of a program invocation (aka instruction).
// Contains all required information to independently replay an instruction.
// Also includes partial transaction context.
message InstrContext {
  // The address of the program invoked.  (32 bytes)
  bytes program_id = 1;

  // The address of the program loader if the invoked program is a user
  // deployed program.
  bytes loader_id = 2;

  // Account state accessed by the instruction.  This may include
  // indirect accesses like sysvars.
  repeated AcctState accounts = 3;

  // Account access list for this instruction (refers to above accounts list)
  repeated InstrAcct instr_accounts = 4;

  // The input data passed to program execution.
  bytes data = 5;

  uint64 cu_avail = 6;

  TxnContext txn_context = 7;
  SlotContext slot_context = 8;
  EpochContext epoch_context = 9;
}

// The results of executing an InstrContext.
message InstrEffects {
  // result is zero if the instruction executed succesfully.
  // Otherwise, a non-zero error code.  Error codes are not relevant to
  // consensus.
  int32 result = 1;

  // Some error cases additionally have a custom error code.  Unlike
  // the expected_result, this is stable across clients.
  uint32 custom_err = 2;

  // Copies of accounts that were changed.  May be in an arbitrary
  // order.  The pubkey of each account is unique in this list.  Each
  // account address modified here must also be in the
  // InstrContext.
  repeated AcctState modified_accounts = 3;

  uint64 cu_avail = 4;

  bytes return_data = 5;
}

// An instruction processing test fixture.
message InstrFixture {
  InstrContext input = 1;
  InstrEffects output = 2;
}

message ELFBinary {
    bytes data = 1;
}

// Wrapper for the ELF binary and the features that the loader should use
// Note that we currently hardcode the features to be used by the loader,
// so features isn't actually used yet.
message ELFLoaderCtx {
    ELFBinary elf = 1;
    FeatureSet features = 2;
}

// Captures the results of a elf binary load.
// Structurally similar to fd_sbpf_program_t
message ELFLoaderEffects {
    bytes rodata = 1;
    uint64 rodata_sz = 2;

    // bytes text = 3; // not needed, just points to a region in rodata
    uint64 text_cnt = 4;
    uint64 text_off = 5;

    uint64 entry_pc = 6;

    repeated uint64 calldests = 7;
}

message ELFLoaderFixture {
    ELFLoaderCtx input = 1;
    ELFLoaderEffects output = 2;
}

//
// Syscalls
//

// Describes an input data region. Agave's memory mapping sets up a series of
// memory mapped regions, which combine to make the input data region.
message InputDataRegion {
  // Offset from the start of the input data segment (0x400000000)
  uint32 offset = 1;
  // Content of the memory region
  bytes content = 2;
  // If the memory region is writable or not
  bool is_writable = 3;
}

// TODO: investigate make this an overlay not a context, and let each target set up the
// vm context from the instruction context.

// Information sufficient to allow the fuzzer to generate a fd_vm_t context for
// execution inside the VM (excluding the instruction context).
//
// TODO: this currently only includes fields necessary for executing syscalls,
// executing sBPF code will require the rest of the fields in fd_vm_t to be set.
message VmContext {
    // Maximum heap size in bytes
    uint64 heap_max = 1;

    // Program read-only data
    bytes rodata = 2;
    // Offset of the text section from the start of the program rodata segment
    // (0x100000000)
    uint64 rodata_text_section_offset = 3;
    // Length of the text section in the program rodata region, in bytes.
    uint64 rodata_text_section_length = 4;

    // The input data regions
    repeated InputDataRegion input_data_regions = 5;

    // Registers
    uint64 r0 = 6;
    uint64 r1 = 7;
    uint64 r2 = 8;
    uint64 r3 = 9;
    uint64 r4 = 10;
    uint64 r5 = 11;
    uint64 r6 = 12;
    uint64 r7 = 13;
    uint64 r8 = 14;
    uint64 r9 = 15;
    uint64 r10 = 16;
    uint64 r11 = 17;

    // TODO: extend this to support the rest of the vm fields, needed for
    // full coverage of the vm interpreter code.
}

// TODO: use structured types for syscall implementation to improve fuzz coverage
// We should seperate out the engine protos from the target protos to make implementation
// of the targets easier.

// TODO: for CPI syscalls, consider generating effects in protobufs vs full cpi
// execution

// A single invocation of a syscall
message SyscallInvocation {
    // The sBPF function name of the syscall
    bytes function_name = 1;
    // The initial portion of the heap, for example to store syscall inputs
    bytes heap_prefix = 2;
}

// Execution context for a VM Syscall execution.
message SyscallContext {
    VmContext vm_ctx = 1;
    // TODO: InflightInstruction - contain temporary fields that live for the duration of an instructions execution, and is needed if we have overhanging context from a previous instruction
    InstrContext instr_ctx = 2;
    SyscallInvocation syscall_invocation = 3;
}

// The effects of executing a SyscallContext.
message SyscallEffects {
    // EBPF error code, if the invocation was unsuccessful
    int64 error = 1;

    // Registers
    uint64 r0 = 2; // Result of a successful execution

    // CU's remaining
    uint64 cu_avail = 3;

    // Memory regions
    bytes heap = 4;
    bytes stack = 5;
    bytes inputdata = 6;

    // Current number of stack frames pushed
    uint64 frame_count = 7;

    // Syscall log
    bytes log = 8;
}

// A syscall processing test fixture.
message SyscallFixture {
  SyscallContext input = 1;
  SyscallEffects output = 2;
}
